internal/ota-api/dto.go

```diff
@@ -155,14 +158,58 @@ func (r OtaStatusDetail) String() string {
 		t = table.New()
 		t.SetHeader("Time", "Status", "Detail")
 		for _, s := range r.Details {
-			t.AddRow(formatHumanReadableTs(s.Timestamp), upperCaseFirst(s.State), s.StateData)
+			stateData := formatStateData(s.State, s.StateData, r.FirmwareSize, hasReachedFlashState(r.Details))
+			t.AddRow(formatHumanReadableTs(s.Timestamp), upperCaseFirst(s.State), stateData)
 		}
 		output += "\nDetails:\n" + t.Render()
 	}
 
 	return output
 }
 
+func hasReachedFlashState(states []State) bool {
+	for _, s := range states {
+		if s.State == "flash" {
+			return true
+		}
+	}
+	return false
+}
+
+func formatStateData(state, data string, firmware_size *int64, hasReceivedFlashState bool) string {
+	if data == "" {
+		return ""
+	}
+	if state == "fetch" {
+		// This is the state 'fetch' of OTA progress. This contains a number that represents the number of bytes fetched
+		if hasReceivedFlashState {
+			return buildSimpleProgressBar(float64(100))
+		}
+		if strings.ToLower(data) == "unknown" {
+			return ""
+		}
+		actualDownloadedData, err := strconv.Atoi(data)
+		if err != nil || firmware_size == nil || actualDownloadedData <= 0 || *firmware_size <= 0 { // Sanitize and avoid division by zero
+			return data
+		}
+		percentage := (float64(actualDownloadedData) / float64(*firmware_size)) * 100
+		return buildSimpleProgressBar(percentage)
+	}
+	if strings.ToLower(data) == "unknown" {
+		return ""
+	}
+	return data
+}
+
+func buildSimpleProgressBar(progress float64) string {
+	progressInt := int(progress) / 10
+	bar := "["
+	bar = bar + strings.Repeat("=", progressInt)
+	bar = bar + strings.Repeat(" ", 10-progressInt)
+	bar = bar + "] " + strconv.FormatFloat(progress, 'f', 2, 64) + "%"
+	return bar
```

Comment:
probably not an big perfomarmance hit, but usually is better to use strings builder to avoid lots of memory allocations. Here a possible version
```suggestion
    progressInt := int(progress) / 10

    var builder strings.Builder
    builder.WriteByte('[')
    builder.WriteString(strings.Repeat("=", progressInt))
    builder.WriteString(strings.Repeat(" ", 10-progressInt))
    builder.WriteByte(']')
    fmt.Fprintf(&builder, " %.2f%%", progress)

    return builder.String()
```

Reply:
it's a cli, so allocation here is not a big deal. Btw, moved to builder approach